### PR TITLE
Support caching for STS server

### PR DESCRIPTION
**Pull request by JimmyCYJ**
Please provide a description for what this PR is for.
Add fetched access token into cache. If the remaining life time is larger than grace period (default: 5 min), STS server returns cached token.

And to help us figure out who should review this PR, please 
put an X in all the areas that this PR affects.

[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[ ] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[x] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

https://github.com/istio/istio/issues/20133

### CODE REVIEW

File: security/pkg/stsservice/sts.go [hunk 1]
```diff
@@ -94,9 +96,10 @@ type TokenManager interface {
 
 // TokenInfo stores token information maintained at TokenManager.
 type TokenInfo struct {
-	TokenType  string `json:"token_type"`
-	IssueTime  string `json:"issue_time"`
-	ExpireTime string `json:"expire_time"`
+	TokenType  string    `json:"token_type"`
+	IssueTime  time.Time `json:"issue_time"`
+	ExpireTime time.Time `json:"expire_time"`
+	Token      string    `json:"token"`
```
Comment (liminw) [inline security/pkg/stsservice/sts.go]: Is tokeinfo printed out only at debug time?

File: security/pkg/stsservice/tokenmanager/google/tokenexchangeplugin.go [hunk 2]
```diff
@@ -135,7 +167,18 @@ func (p *Plugin) constructFederatedTokenRequest(parameters stsservice.StsRequest
 	req, _ := http.NewRequest("POST", federatedTokenEndpoint, bytes.NewBuffer(jsonQuery))
 	req.Header.Set("Content-Type", contentType)
 	if pluginLog.DebugEnabled() {
-		reqDump, _ := httputil.DumpRequest(req, false)
+		dQuery := map[string]string{
+			"audience":           aud,
+			"grantType":          parameters.GrantType,
+			"requestedTokenType": tokenType,
+			"subjectTokenType":   parameters.SubjectTokenType,
+			"subjectToken":       "redacted",
```
Comment (liminw) [inline security/pkg/stsservice/tokenmanager/google/tokenexchangeplugin.go]: "redacted". Is this intended?

File: security/pkg/stsservice/tokenmanager/google/mock/mockserver.go [hunk 3]
```diff
@@ -329,11 +329,9 @@ func (ms *AuthorizationServer) getAccessToken(w http.ResponseWriter, req *http.R
 		w.WriteHeader(http.StatusInternalServerError)
 		return
 	}
-	tokenLife.Format(time.RFC3339)
-	lifeJSON, _ := tokenLife.MarshalJSON()
 	resp := accessTokenResponse{
 		AccessToken: token,
-		ExpireTime:  string(lifeJSON),
+		ExpireTime:  tokenLife.Format(time.RFC3339Nano),
```
Comment (howardjohn) [inline security/pkg/stsservice/tokenmanager/google/mock/mockserver.go]: isn't this changing from json? is that expected?

File: security/pkg/stsservice/tokenmanager/google/tokenexchangeplugin.go [hunk 4]
```diff
@@ -135,7 +167,18 @@ func (p *Plugin) constructFederatedTokenRequest(parameters stsservice.StsRequest
...
+			"scope":              reqScope,
+		}
+		dJSONQuery, _ := json.Marshal(dQuery)
+		dReq, _ := http.NewRequest("POST", federatedTokenEndpoint, bytes.NewBuffer(dJSONQuery))
```
Comment (howardjohn) [inline security/pkg/stsservice/tokenmanager/google/tokenexchangeplugin.go]: why ignore all the errors here?

File: security/pkg/stsservice/tokenmanager/sts_test.go [hunk 5]
```diff
@@ -161,16 +198,22 @@ func sendHTTPRequestWithRetry(client *http.Client, req *http.Request) (resp *htt
 	return resp, err
 }
 
+type testSetUp struct {
+	enableCache        bool
+	enableDynamicToken bool
```
Comment (liminw) [inline security/pkg/stsservice/tokenmanager/sts_test.go]: What does enableDynamicToken mean?

File: security/pkg/stsservice/tokenmanager/google/tokenexchangeplugin.go [hunk 6]
```diff
@@ -313,11 +365,15 @@ func (p *Plugin) generateSTSResp(atResp *accessTokenResponse) ([]byte, error) {
...
 	stsRespParam := stsservice.StsResponseParameters{
-		AccessToken:     atResp.AccessToken,
+		AccessToken:     token,
 		IssuedTokenType: tokenType,
 		TokenType:       "Bearer",
-		ExpiresIn:       expireInSec,
+		ExpiresIn:       expire,
 	}
 	statusJSON, err := json.MarshalIndent(stsRespParam, "", " ")
 	if pluginLog.DebugEnabled() {
```
Comment (liminw) [inline security/pkg/stsservice/tokenmanager/google/tokenexchangeplugin.go]: "redacted"?

File: security/pkg/stsservice/tokenmanager/google/mock/mockserver.go (hunk 3 above)
Comment (JimmyCYJ) [reply security/pkg/stsservice/tokenmanager/google/mock/mockserver.go]: I did search for that. One example of using JSON is like this
https://stackoverflow.com/a/23695774/8922031.
If we want to specify a time format when marshalling time, we need to create our helper function.
Please let me know if there is a better option.

File: security/pkg/stsservice/tokenmanager/google/tokenexchangeplugin.go (hunk 4 above)
Comment (JimmyCYJ) [reply security/pkg/stsservice/tokenmanager/google/tokenexchangeplugin.go]: This part is to dump the request for troubleshooting purposes, dJSONQuery and dReq are not sent to the endpoint.

File: security/pkg/stsservice/tokenmanager/google/tokenexchangeplugin.go (hunk 6 above)
Comment (JimmyCYJ) [reply security/pkg/stsservice/tokenmanager/google/tokenexchangeplugin.go]: If expire time is redacted, then the debug mode does not provide useful information.

File: security/pkg/stsservice/sts.go (hunk 1 above)
Comment (JimmyCYJ) [reply security/pkg/stsservice/sts.go]: This is not used for debug, when we turn on debug, this is not printed. Only used for caching.

File: security/pkg/stsservice/tokenmanager/google/tokenexchangeplugin.go (hunk 2 above)
Comment (JimmyCYJ) [reply security/pkg/stsservice/tokenmanager/google/tokenexchangeplugin.go]: This is intended. This is only used in debug log.

File: security/pkg/stsservice/tokenmanager/sts_test.go (hunk 5 above)
Comment (JimmyCYJ) [reply security/pkg/stsservice/tokenmanager/sts_test.go]: So the fake token has a static prefix and timestamp, if two STS requests get different token, then the token from second response is not served from cache.

**Pull request merged**